internal/interpreter/dynamic_component.rs

```diff
@@ -1834,36 +1880,53 @@ impl<'a, 'id> InstanceRef<'a, 'id> {
     }
 
     pub fn window_adapter(&self) -> Rc<dyn WindowAdapter> {
+        let root_weak = vtable::VWeak::into_dyn(self.root_weak().clone());
         let root = self.root_weak().upgrade().unwrap();
         generativity::make_guard!(guard);
         let comp = root.unerase(guard);
         Self::get_or_init_window_adapter_ref(
             &comp.component_type,
+            root_weak,
             comp.instance.as_pin_ref().get_ref(),
         )
         .unwrap()
         .clone()
     }
 
-    // Call this only on root components!
     pub fn get_or_init_window_adapter_ref<'b, 'id2>(
         component_type: &'b ComponentDescription<'id2>,
+        root_weak: ComponentWeak,
         instance: &'b Instance<'id2>,
     ) -> Result<&'b Rc<dyn WindowAdapter>, PlatformError> {
+        // We are the actual root: Generate and store a window_adapter if necessary
         component_type.window_adapter_offset.apply(instance).get_or_try_init(|| {
-            let extra_data = component_type.extra_data_offset.apply(instance);
-            let window_adapter = i_slint_backend_selector::with_platform(|_b| {
-                #[cfg(not(target_arch = "wasm32"))]
-                return _b.create_window_adapter();
-                #[cfg(target_arch = "wasm32")]
-                i_slint_backend_winit::create_gl_window_with_canvas_id(
-                    extra_data.canvas_id.get().map_or("canvas", |s| s.as_str()),
-                )
-            })?;
-            let comp_rc = extra_data.self_weak.get().unwrap().upgrade().unwrap();
-            WindowInner::from_pub(window_adapter.window())
-                .set_component(&vtable::VRc::into_dyn(comp_rc));
-            Ok(window_adapter)
+            let mut parent_node = ItemWeak::default();
+            if let Some(rc) = vtable::VWeak::upgrade(&root_weak) {
+                vtable::VRc::borrow_pin(&rc).as_ref().parent_node(&mut parent_node);
+            }
+
+            if let Some(parent) = parent_node.upgrade() {
+                // We are embedded: Get window adapter from our parent
+                let mut result = None;
+                vtable::VRc::borrow_pin(parent.component()).as_ref().window_adapter(&mut result);
```

Comment:
This looks good to me. Isn't the same also needed for maybe_window_adapter() need the same forwarding to the component vtable?

Reply:
Are you sure that it is stored when called from the embedded component? I somehow missed that part.

Reply:
I think the one place where it is needed is in the destructor.

If an item tree was rendered once, it’s crucial that in the destructor maybe_window_adapter returns true. If not (rendered) then we certainly don’t want to create the adapter.

Failing to call unregister_component for a previously rendered item tree will result in memory leaks.

Therefore for an embedded component it’s crucial in the destructor to either be sure that the field you mentioned is populated or that we go back up the tree of components to find the adapter.